File: StriverDP/numberOfSubsets.py
'''
https://www.codingninjas.com/codestudio/problems/number-of-subsets_3952532?source=youtube&campaign=striver_dp_videos&leftPanelTab=0
'''

from os import *
from sys import *
from collections import *
from math import *

from typing import List

def findWays(num: List[int], tar: int) -> int:
    n = len(num)
    dp = [[0]*(tar+1) for i in range(n)]
    for i in range(n):
        dp[i][0] = 1
    try:
        dp[0][num[0]] = 1
    except:
        pass
    for ind in range(n):
        for target in range(tar+1):
            notTake = dp[ind-1][target]
            take = 0
            if(num[ind] <= target):
                take = dp[ind-1][target-num[ind]]
            dp[ind][target] = take + notTake
    
    return dp[n-1][tar]

File: StriverDP/NumberOfSubsets.cpp
/*
https://www.codingninjas.com/codestudio/problems/number-of-subsets_3952532?source=youtube&campaign=striver_dp_videos&leftPanelTab=0
*/
#include <bits/stdc++.h>

using namespace std;

int findWaysUtil(int ind, int target, vector<int>& arr, vector<vector<int>> &dp){
    if(target==0)
        return 1;
    
    if(ind == 0)
        return arr[0] == target;
    
    if(dp[ind][target]!=-1)
        return dp[ind][target];
        
    int notTaken = findWaysUtil(ind-1,target,arr,dp);
    
    int taken = 0;
    if(arr[ind]<=target)
        taken = findWaysUtil(ind-1,target-arr[ind],arr,dp);
        
    return dp[ind][target]= notTaken + taken;
}

int findWays(vector<int> &num, int k){
    int n = num.size();
    vector<vector<int>> dp(n,vector<int>(k+1,-1));
    return findWaysUtil(n-1,k,num,dp);
}
